elastic/apm-server
# Use dedicated overflow buckets for txn and svc aggregation to limit cardinality

## Motivation/summary

This PR introduces dedicated `other` buckets for transaction and service aggregation metrics to limit cardinality. The overflow buckets produce a metric named `overflow_count` that estimates the number of unique transactions/services that were aggregated in these buckets.

The limits for these overflow buckets are derived from `max_transaction_groups` for transaction metrics and `max_groups` for service metrics. These limits, if not explicitly specified, are derived from the memory limit available during runtime using the formula `gb_available*10_000`.

## Checklist

~~- [ ] Documentation has been updated~~

## How to test these changes

1. Run APM-Server with a known memory limit.
2. Check the logs of APM-Server to validate the configuration for `MaxTransactionGroups` and `MaxGroups`. The logs are in the format `MaxTransactionGroups set to xx based on xxgb of memory` and `MaxGroups set to xx based on xxgb of memory`. Assert that these two limits are approximately equal to `gb_available*10_000`.
3. Send a lot of transactions such that the cardinality of 10_000 is exceeded in a period of 1 minute.
4. Assert that metric documents are published with `transaction.name: other`.

## Related issues

Part of #9648
